File: 5-mle/2-langgraph/1-langgraph-basic/graph-advanced/09_observability/tools.py
"""데모용 도구."""
from __future__ import annotations

from datetime import datetime, timedelta, timezone

from langchain_core.tools import tool


@tool
def get_current_time() -> str:
    """현재 시간(KST)을 반환합니다."""
    kst = timezone(timedelta(hours=9))
    return datetime.now(kst).strftime("%Y-%m-%d %H:%M:%S KST")


@tool
def calculate(expression: str) -> str:
    """수학 식을 안전하게 계산합니다. 숫자와 +-*/(). 만 허용."""
    allowed = set("0123456789+-*/.() ")
    if not all(c in allowed for c in expression):
        return "Error: Only numeric expressions allowed."
    try:
        return str(eval(expression))  # noqa: S307
    except Exception as e:
        return f"Error: {e}"


TOOLS = [get_current_time, calculate]

__all__ = ["TOOLS"]

File: 5-mle/2-langgraph-js/apps/bff/src/langgraph-examples/graph-advanced/09-observability/tools.ts
import { tool } from "@langchain/core/tools";
import { z } from "zod";

export const getCurrentTime = tool(
  () => {
    const now = new Date();
    const kst = new Intl.DateTimeFormat("sv-SE", {
      timeZone: "Asia/Seoul",
      year: "numeric",
      month: "2-digit",
      day: "2-digit",
      hour: "2-digit",
      minute: "2-digit",
      second: "2-digit",
      hour12: false
    }).format(now);
    return `${kst.replace("T", " ")} KST`;
  },
  {
    name: "get_current_time",
    description: "현재 시간(KST)을 반환합니다.",
    schema: z.object({})
  }
);

export const calculate = tool(
  ({ expression }) => {
    const allowed = /^[0-9+\-*/.() ]+$/;
    if (!allowed.test(expression)) {
      return "Error: Only numeric expressions allowed.";
    }

    try {
      return String(Function(`"use strict"; return (${expression});`)());
    } catch (error) {
      return `Error: ${error instanceof Error ? error.message : String(error)}`;
    }
  },
  {
    name: "calculate",
    description: "숫자와 +-*/(). 만 포함된 수학 식을 계산합니다.",
    schema: z.object({
      expression: z.string().describe("계산할 수식")
    })
  }
);

export const TOOLS = [getCurrentTime, calculate];
